python_modules/dagster-test/dagster_test/toys/external_system_assets/__init__.py

```diff
@@ -0,0 +1,102 @@
+"""Models a case where Dagster is used purely for orchestration and not data transformation.
+
+In this scenario, there is an asset graph defined in some external system. An asset factory is used
+to build Dagster `AssetsDefinition` objects corresponding to the nodes in the external system's
+asset graph. The factory queries the external system for dependency and code version info.
+
+Dagster forwards materialization requests to the external system. We pass an asset key as a string
+together with serialized provenance information for the last materialization of the asset on record.
+The backend determines what to do with this information-- if it determines that the provenance of
+the last materialization is not stale (different from the current state of the asset), then it will
+simply use the existing stored value of the asset. Otherwise, it will recompute the asset and store
+the result.
+
+The materialization endpoint of the external system returns a data version string and a flag
+indicating whether a memoized value was used. This information is passed to the Dagster framework by
+returning a `Nothing` `Output`.
+"""
+import json
+import warnings
+from typing import cast
+
+from dagster import OpExecutionContext, observable_source_asset
+from dagster._core.definitions.assets import AssetsDefinition
+from dagster._core.definitions.data_version import DataProvenance, DataVersion
+from dagster._core.definitions.definitions_class import Definitions
+from dagster._core.definitions.events import AssetKey, Output
+from dagster._core.definitions.input import In
+from dagster._core.definitions.op_definition import OpDefinition
+from dagster._core.definitions.output import Out
+from dagster._core.types.dagster_type import Nothing
+from dagster._utils.backcompat import ExperimentalWarning
+
+from .external_system import ExternalSystem
+
+warnings.filterwarnings("ignore", category=ExperimentalWarning)
+
+EXTERNAL_SYSTEM = ExternalSystem()
+
+
+def external_asset(key: AssetKey):
```

Comment:
Any reason why this doesn't just use `multi_asset`?

Reply:
That's a good question. I haven't fully thought it through, but a constraint on a multi-asset is that the same op is used for all assets. I would need to add additional logic into this master op to handle the selection and yield appropriate outputs. Seems like additional complexity without benefit?

If this is what our stakeholders are using or what you'd like to recommend to them then I can change it, but to me atm it seems cleaner to keep it as is.